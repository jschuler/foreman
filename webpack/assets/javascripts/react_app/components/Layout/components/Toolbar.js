import React from 'react';
import {
  DataToolbar,
  DataToolbarContent,
  DataToolbarGroup,
  DataToolbarItem,
} from '@patternfly/react-core';
import TaxonomySwitcher from './TaxonomySwitcher';
import UserDropdowns from './UserDropdowns';
import NotificationContainer from '../../notifications';
import ImpersonateIcon from './ImpersonateIcon';
import { layoutPropTypes, layoutDefaultProps } from '../LayoutHelper';

const Toolbar = ({
  data,
  currentLocation,
  changeLocation,
  currentOrganization,
  changeOrganization,
  isLoading,
  changeActiveMenu,
  ...props
}) => (
  <DataToolbar id="data-toolbar" {...props}>
    <DataToolbarContent style={{ width: '100%' }}>
      <DataToolbarGroup
        breakpointMods={[
          { modifier: 'space-items-sm' },
          { modifier: 'hidden' },
          { modifier: 'visible', breakpoint: 'xl' },
        ]}
      >
        <TaxonomySwitcher
          taxonomiesBool={data.taxonomies}
          currentLocation={currentLocation}
          locations={
            data.taxonomies.locations ? data.locations.available_locations : []
          }
          onLocationClick={changeLocation}
          currentOrganization={currentOrganization}
          organizations={
            data.taxonomies.organizations
              ? data.orgs.available_organizations
              : []
          }
          onOrgClick={changeOrganization}
          isLoading={isLoading}
        />
      </DataToolbarGroup>
      <DataToolbarGroup
        breakpointMods={[
          { modifier: 'space-items-md' },
          { modifier: 'align-right' },
          { modifier: 'hidden' },
          { modifier: 'visible', breakpoint: 'md' },
        ]}
      >
        <DataToolbarItem>
          <NotificationContainer data={{ url: data.notification_url }} />
        </DataToolbarItem>
        {data.user.impersonated_by && (
          <DataToolbarItem>
            <ImpersonateIcon
              stopImpersonationUrl={data.stop_impersonation_url}
            />
          </DataToolbarItem>
        )}
        <DataToolbarItem>
          <UserDropdowns
            notificationUrl={data.notification_url}
            user={data.user}
            changeActiveMenu={changeActiveMenu}
            stopImpersonationUrl={data.stop_impersonation_url}
          />
        </DataToolbarItem>
      </DataToolbarGroup>
    </DataToolbarContent>
  </DataToolbar>
);
Toolbar.propTypes = {
  data: layoutPropTypes.data,
  currentLocation: layoutPropTypes.currentLocation,
  changeLocation: layoutPropTypes.changeLocation,
  currentOrganization: layoutPropTypes.currentOrganization,
  changeOrganization: layoutPropTypes.changeOrganization,
  isLoading: layoutPropTypes.isLoading,
  changeActiveMenu: layoutPropTypes.changeActiveMenu,
};
Toolbar.defaultProps = {
  data: layoutDefaultProps.data,
  currentLocation: layoutDefaultProps.currentLocation,
  changeLocation: layoutDefaultProps.changeLocation,
  currentOrganization: layoutDefaultProps.currentOrganization,
  changeOrganization: layoutDefaultProps.changeOrganization,
  isLoading: layoutDefaultProps.isLoading,
  changeActiveMenu: layoutDefaultProps.changeActiveMenu,
};
export default Toolbar;
