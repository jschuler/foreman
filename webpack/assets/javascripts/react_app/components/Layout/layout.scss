@import '../../common/colors.scss';

.pf-c-page.foreman-theme {
  .pf-c-page__header {
    .pf-c-data-toolbar {
      background-color: transparent;
    }
  }

  .pf-c-page__sidebar.pf-m-dark {
    background-color: $nav-pf-vertical-secondary-active-bg-color;

    .pf-c-nav.pf-m-dark > .pf-c-nav__list {
      .pf-c-nav__item.pf-m-current,
      .pf-c-nav__item > .pf-c-nav__link.pf-m-current {
        background-color: $nav-pf-vertical-active-bg-color;
      }

      .pf-c-nav__item.pf-m-current > .pf-c-nav__link[data-component=pf-nav-expandable] {
        color: $nav-pf-vertical-active-icon-color;
      }

      .pf-c-nav__link {
        color: $nav-pf-vertical-icon-color;
      }

      .pf-c-nav__simple-list .pf-c-nav__link {
        &.pf-m-current,
        &.pf-m-hover,
        &:hover {
          background-color: $nav-pf-vertical-active-bg-color;
          color: $nav-pf-vertical-active-icon-color;
        }
      }
    }
  }
}

#account_menu {
  outline: none;
}

#vertical-spinner {
  margin-top: 15px;
  margin-left: 5px;
}
