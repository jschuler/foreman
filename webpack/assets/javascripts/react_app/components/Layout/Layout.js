import React from 'react';
import {
  Brand,
  Page,
  PageSection,
  PageSectionVariants,
  PageSidebar,
  SkipToContent,
} from '@patternfly/react-core';
// eslint-disable-next-line
import globalBreakpointMd from '@patternfly/react-tokens/dist/js/global_breakpoint_md';
import CustomPageHeader from './components/CustomPageHeader';
import {
  getActive,
  getCurrentPath,
  layoutPropTypes,
  layoutDefaultProps,
} from './LayoutHelper';
import VerticalNav from './components/VerticalNav';
import Toolbar from './components/Toolbar';
import './layout.scss';

class Layout extends React.Component {
  componentDidMount() {
    const {
      items,
      data,
      fetchMenuItems,
      changeLocation,
      currentLocation,
      changeOrganization,
      currentOrganization,
      changeActiveMenu,
      activeMenu,
    } = this.props;
    if (items.length === 0) fetchMenuItems(data);

    const activeURLMenu = getActive(data.menu, getCurrentPath());
    if (activeMenu !== activeURLMenu.title) {
      changeActiveMenu(activeURLMenu);
    }

    if (
      data.taxonomies.locations &&
      !!data.locations.current_location &&
      currentLocation !== data.locations.current_location
    ) {
      const initialLocTitle = data.locations.current_location;
      const initialLocId = data.locations.available_locations.find(
        loc => loc.title === initialLocTitle
      ).id;
      changeLocation({ title: initialLocTitle, id: initialLocId });
    }

    if (
      data.taxonomies.organizations &&
      !!data.orgs.current_org &&
      currentOrganization !== data.orgs.current_org
    ) {
      const initialOrgTitle = data.orgs.current_org;
      const initialOrgId = data.orgs.available_organizations.find(
        org => org.title === initialOrgTitle
      ).id;
      changeOrganization({ title: initialOrgTitle, id: initialOrgId });
    }
  }

  render() {
    const {
      items,
      data,
      isLoading,
      changeActiveMenu,
      changeOrganization,
      changeLocation,
      currentOrganization,
      currentLocation,
      activeMenu,
      children,
      history,
    } = this.props;

    const afterNavToggle = navOpen => {
      const mobileView =
        window.innerWidth < Number.parseInt(globalBreakpointMd.value, 10);
      if (mobileView) {
        // ignore for mobile view
        return;
      }
      // toggles a class in the body tag, so that the main #rails-app-content container can have the appropriate width
      if (navOpen) {
        document.body.classList.add('pf-m-expanded');
      } else {
        document.body.classList.remove('pf-m-expanded');
      }
    };
    const header = (
      <CustomPageHeader
        logo={
          <Brand
            src={data.logo}
            alt={data.brand}
            style={{ marginTop: '8px' }}
          />
        }
        logoProps={{ href: data.root }}
        showNavToggle
        contextSelector={
          <Toolbar
            data={data}
            currentLocation={currentLocation}
            changeLocation={changeLocation}
            currentOrganization={currentOrganization}
            changeOrganization={changeOrganization}
            isLoading={isLoading}
            changeActiveMenu={changeActiveMenu}
          />
        }
        afterNavToggle={afterNavToggle}
        className="navbar-pf-vertical"
      />
    );
    const sidebar = (
      <PageSidebar
        nav={
          <VerticalNav
            items={items}
            activeMenu={activeMenu}
            changeActiveMenu={changeActiveMenu}
            history={history}
          />
        }
        theme="dark"
      />
    );
    const pageId = 'main';
    const pageSkipToContent = (
      <SkipToContent href={`#${pageId}`}>Skip to content</SkipToContent>
    );

    return (
      <Page
        header={header}
        sidebar={sidebar}
        isManagedSidebar
        skipToContent={pageSkipToContent}
        mainContainerId={pageId}
        className="foreman-theme"
      >
        <PageSection variant={PageSectionVariants.light}>
          {children}
        </PageSection>
      </Page>
    );
  }
}

Layout.propTypes = {
  ...layoutPropTypes,
};

Layout.defaultProps = {
  ...layoutDefaultProps,
};

export default Layout;
