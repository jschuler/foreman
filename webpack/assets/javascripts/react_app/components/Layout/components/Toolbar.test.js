import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';
import { hasTaxonomiesMock } from '../Layout.fixtures';

import Toolbar from './Toolbar';

const fixtures = {
  'render Toolbar': {
    data: hasTaxonomiesMock.data,
    currentLocation: hasTaxonomiesMock.currentLocation,
    changeLocation: () => {},
    currentOrganization: hasTaxonomiesMock.currentOrganization,
    changeOrganization: () => {},
    isLoading: false,
    changeActiveMenu: () => {},
  },
};

describe('Toolbar', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(Toolbar, fixtures));
});
