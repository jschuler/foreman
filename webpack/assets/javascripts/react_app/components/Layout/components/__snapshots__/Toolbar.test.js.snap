// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar rendering render Toolbar 1`] = `
<DataToolbar
  id="data-toolbar"
>
  <DataToolbarContent
    breakpointMods={Array []}
    isExpanded={false}
    showClearFiltersButton={false}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <ForwardRef
      breakpointMods={
        Array [
          Object {
            "modifier": "space-items-sm",
          },
          Object {
            "modifier": "hidden",
          },
          Object {
            "breakpoint": "xl",
            "modifier": "visible",
          },
        ]
      }
    >
      <TaxonomySwitcher
        currentLocation="london"
        currentOrganization="org1"
        isLoading={false}
        locations={
          Array [
            Object {
              "href": "/locations/1-yaml/select",
              "id": 1,
              "title": "yaml",
            },
            Object {
              "href": "/locations/2-london/select",
              "id": 2,
              "title": "london",
            },
            Object {
              "href": "/locations/3-norway/select",
              "id": 3,
              "title": "norway",
            },
          ]
        }
        onLocationClick={[Function]}
        onOrgClick={[Function]}
        organizations={
          Array [
            Object {
              "href": "/organizations/1-org1/select",
              "id": 1,
              "title": "org1",
            },
            Object {
              "href": "/organizations/2-org2/select",
              "id": 2,
              "title": "org2",
            },
          ]
        }
        taxonomiesBool={
          Object {
            "locations": true,
            "organizations": true,
          }
        }
      />
    </ForwardRef>
    <ForwardRef
      breakpointMods={
        Array [
          Object {
            "modifier": "space-items-md",
          },
          Object {
            "modifier": "align-right",
          },
          Object {
            "modifier": "hidden",
          },
          Object {
            "breakpoint": "md",
            "modifier": "visible",
          },
        ]
      }
    >
      <DataToolbarItem>
        <Connect(OnClickOutside(notificationContainer))
          data={
            Object {
              "url": "/notification_recipients",
            }
          }
        />
      </DataToolbarItem>
      <DataToolbarItem>
        <UserDropdowns
          changeActiveMenu={[Function]}
          className=""
          notificationUrl="/notification_recipients"
          stopImpersonationUrl="/users/stop_impersonation"
          user={
            Object {
              "current_user": Object {
                "user": Object {
                  "firstname": "Admin",
                  "id": 4,
                  "lastname": "User",
                  "login": "admin",
                  "name": "Admin User",
                },
              },
              "user_dropdown": Array [
                Object {
                  "children": Array [
                    Object {
                      "href": "/Aa",
                      "isDivider": false,
                      "onClick": [MockFunction],
                      "title": "Aa",
                    },
                    Object {
                      "href": "/Cc",
                      "isDivider": false,
                      "onClick": [MockFunction],
                      "title": "Cc",
                    },
                  ],
                  "icon": "fa fa-user",
                  "name": "User",
                  "type": "sub_menu",
                },
              ],
            }
          }
        />
      </DataToolbarItem>
    </ForwardRef>
  </DataToolbarContent>
</DataToolbar>
`;
